elytron/WFCORE-3947-ocsp.adoc

```diff
@@ -0,0 +1,67 @@
+= WFCORE-3947 OCSP support in Elytron SSL truststore
+:author:            Jan Kalina
+:email:             jkalina@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          elytron
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Elytron provides trustmanager usable for SSL contexts used by other parts of WildFly.
+It supports certificate revocations using certificate revocations list (CRL), but support of Online Certificate Status Protocol (OCSP) is required too.
+This is required especially for client certificates validation.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/EAP7-715[EAP7-715]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFCORE-3947[WFCORE-3947]
+
+=== Dev Contacts
+
+* mailto:jkalina@redhat.com[Jan Kalina]
+
+=== QE Contacts
+
+* mailto:jstourac@redhat.com[Jan Stourac]
+
+=== Affected Projects or Components
+
+* WildFly Core
+* WildFly Elytron
+
+//=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* OCSP will be supported as an alternative to CRL in Elytron subsystem trustmanager (for client certificates verification).
```

Comment:
Do we want to enforce exclusive use here? I mean - only CRL or OCSP can be configured and used exclusively at one moment?

Reply:
We can support both in theory, but we need to define how to combine them:
* a) should be CRL used only when OCSP is unavailable?
* b) should be every certificate checked against both of them?

Reply:
To be honest, I have no idea whether it is common requirement to have both checks available and configured or not.

If we do want to support it, then I have found that some tool from Citrix is able to work like that [1] - both CRL and OCSP can be configured. Here is their decision/priority table:

>* If OCSPcheck parameter is set as mandatory, CRLcheck parameter will not be effective even if it is optionally enabled internally.
>* If CRL check is mandatory, then CRL check will be effective even if OCSP is configured.
>* If OCSP and CRL check are optional, then OCSP configuration is given first preference and then CRL check applies.
>* Both OCSP and CRL check cannot be set as mandatory.

We could probably mimic this somehow...

Also, if both CRL and OCSP are configured as optional, we should perform check request for both of them right away so we have result of the querry ASAP for both type of checks, not waiting e.g. timeout of the OCSP to respond...

[1] https://support.citrix.com/article/CTX138304

Reply:
* OCSP only is implemented
* CRL only is implemented
* OCSP with CRL fallback is not implemented yet 
@darranl do we want to support OCSP with CRL fallback?

Reply:
Unless we have a strong argument it is required I would suggest we pull additional strategies into their own RFEs.

Once we have the core feature in place it will be a lot quicker to make minor enhancements.